src/server/scripts/EasternKingdoms/SunwellPlateau/boss_kalecgos.cpp

```diff
@@ -409,100 +354,76 @@ struct boss_kalec : public ScriptedAI
-
-        switch (events.ExecuteEvent())
-        {
-        case EVENT_CHECK_HEALTH:
-            if (me->HealthBelowPct(50))
-            {
-                Talk(SAY_GOOD_NEAR_DEATH);
-                break;
-            }
-            events.ScheduleEvent(EVENT_CHECK_HEALTH, 1000);
-            break;
-        case EVENT_CHECK_HEALTH2:
-            if (me->HealthBelowPct(10))
-            {
-                Talk(SAY_GOOD_NEAR_DEATH2);
-                break;
-            }
-            events.ScheduleEvent(EVENT_CHECK_HEALTH2, 1000);
-            break;
-        case EVENT_SPELL_REVITALIZE:
-            me->CastSpell(me, SPELL_REVITALIZE, false);
-            events.ScheduleEvent(EVENT_SPELL_REVITALIZE, 10000);
-            break;
-        case EVENT_SPELL_HEROIC_STRIKE:
-            me->CastSpell(me->GetVictim(), SPELL_HEROIC_STRIKE, false);
-            events.ScheduleEvent(EVENT_SPELL_HEROIC_STRIKE, 5000);
-            break;
-        }
-
-        DoMeleeAttackIfReady();
+        scheduler.Update(diff,
+            std::bind(&BossAI::DoMeleeAttackIfReady, this));
     }
 };
 
 struct boss_sathrovarr : public ScriptedAI
 {
     boss_sathrovarr(Creature* creature) : ScriptedAI(creature)
     {
-        instance = creature->GetInstanceScript();
+        _instance = creature->GetInstanceScript();
+        SetInvincibility(true);
     }
 
-    InstanceScript* instance;
-    EventMap events;
-
     bool CanAIAttack(Unit const* target) const override
     {
         return target->GetPositionZ() < 50.0f;
     }
 
     void Reset() override
     {
-        events.Reset();
-        me->CastSpell(me, SPELL_DEMONIC_VISUAL, true);
-        me->CastSpell(me, SPELL_SPECTRAL_INVISIBILITY, true);
-
-        events.ScheduleEvent(EVENT_SHADOW_BOLT, 7000);
-        events.ScheduleEvent(EVENT_AGONY_CURSE, 20000);
-        events.ScheduleEvent(EVENT_CORRUPTION_STRIKE, 13000);
-        events.ScheduleEvent(EVENT_CHECK_HEALTH, 1000);
-        events.ScheduleEvent(EVENT_CHECK_HEALTH2, 1000);
+        DoCastSelf(SPELL_DEMONIC_VISUAL, true);
+        DoCastSelf(SPELL_SPECTRAL_INVISIBILITY, true);
     }
 
     void JustEngagedWith(Unit* /*who*/) override
     {
         Talk(SAY_SATH_AGGRO);
-    }
 
-    void DamageTaken(Unit* who, uint32& damage, DamageEffectType, SpellSchoolMask) override
-    {
-        if (damage >= me->GetHealth() && who != me)
-            damage = 0;
+        ScheduleTimedEvent(7s, [&] {
+            if (roll_chance_i(20))
+                Talk(SAY_SATH_SPELL1);
+            DoCastVictim(SPELL_SHADOW_BOLT);
+        }, 9s);
+
+        ScheduleTimedEvent(20s, [&] {
+            me->CastCustomSpell(SPELL_CURSE_OF_BOUNDLESS_AGONY, SPELLVALUE_MAX_TARGETS, 1, me, false);
+        }, 30s);
+
+        ScheduleTimedEvent(20s, [&] {
+            if (roll_chance_i(20))
+                Talk(SAY_SATH_SPELL2);
+            DoCastVictim(SPELL_CORRUPTION_STRIKE);
+        }, 9s);
+
+        scheduler.Schedule(1s, [this](TaskContext context)
```

Comment:
Same here - why not use ScheduleHealthCheck?
Otherwise these 2 checks could be combined into one event maybe?

Reply:
No because once you’re below 10% you can’t be checking for below 10% anymore, it needs to stop on the first tick.